docs/design/features/byreflike-generics.md

```diff
@@ -0,0 +1,75 @@
+
+- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+- `throw` &ndash; Requires an object reference on the stack so not directly impacted since boxing is not permitted.
+- `stsfld` / `ldsfld` &ndash; Type fields of a ByRefLike parameter cannot be marked `static`.
+- `newarr` / `stelem` / `ldelem` / `ldelema` &ndash; Arrays are not able to contain ByRefLike types.
+    - `newobj` &ndash; For multi-dimensional array construction.
+- `constrained.callvirt` &ndash; This IL sequence must resolve to a method implemented on `object`, or a default interface method.
+- Use of any Generic which doesn’t expect a ByRefLike parameter as a Generic parameter.
+
+## Proposal
+
+A new Attribute API will be defined to indicate which Generic parameters are permissible to be of any type&mdash;including ByRefLike types. The Attribute could be used by the compiler to implement a generic non-constraint.
+
+```csharp
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
+    public class GenericParameterSupportsAnyTypeAttribute : Attribute
+    {
+        /// </summary>
+        /// Constructs an instance with a Generic parameter index of 0.
+        /// </summary>
+        public GenericParameterSupportsAnyTypeAttribute()
+            => ParameterIndex = 0;
+
+        /// </summary>
+        /// Constructs an instance with a Generic parameter index.
+        /// </summary>
+        /// <param name="parameterIndex">Non-negative index of Generic parameter to apply to.</param>
+        public GenericParameterSupportsAnyTypeAttribute(int parameterIndex)
+            => ParameterIndex = parameterIndex;
+
+        /// <summary>
+        /// Generic Parameter Index.
+        /// </summary>
+        public int ParameterIndex { get; set; }
+    }
+}
+```
+
+A new API will be implemented as a JIT intrinsic for determining if a parameter is ByRefLike. This API would represent a check to occur at JIT time code-gen to avoid taking paths that would be invalid for some values of `T`.
+
+```diff
+namespace System
+{
+    public abstract partial class Type
+    {
++        [Intrinsic]
++        public static bool IsByRefLike<T>();
+    }
+}
+```
+
+For dispatch to object implemented methods and to default interface methods, the behavior shall be that an `InvalidProgramException` should be thrown. The JIT would insert the following IL at code-gen time.
+
+```
+newobj instance void System.InvalidProgramException::.ctor()
+throw
+```
+
+When boxing due to a constrained call that cannot be made, instead of allocating a normal boxed object, an object of `InvalidBoxedObject` type will be created. It will have implementations of the various overridable object methods which throw `InvalidProgramException`, and interface dispatch shall have a special case for attempting to invoke an interface method on such an object, that will also throw an `InvalidProgramException`.
```

Comment:
Why do we need the `InvalidBoxedObject`? Can't we just throw `InvalidProgram` during the boxing attempt?

Reply:
I think I see:

1. If we throw during boxing, we would throw "Type 'SomeRefLikeType' needs to be boxed to perform this call, but the type is byref-like." (not giving context on the fact what interface/base virtual call we try to make).
2. If we make an InvalidBoxedObject, we would throw "Cannot call method IFoo.SomeMethod on this type because it's byref-like" (not giving context on the type of `this` instead).

1 has less implementation complexity (it may actually already work, modulo the more specific message). 2 gives more detail in a way but also less detail in another way.

Reply:
I was given some additional feedback on this approach. I'm going to simply apply (1) unless we determine something like (2) is needed.